internal/provisioner/objects/contourconfig/contourconfig.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contourconfig
+
+import (
+	"context"
+
+	contour_api_v1alpha1 "github.com/projectcontour/contour/apis/projectcontour/v1alpha1"
+	"github.com/projectcontour/contour/internal/provisioner/labels"
+	"github.com/projectcontour/contour/internal/provisioner/model"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// EnsureContourConfig ensures that a ContourConfiguration exists for the given contour.
+func EnsureContourConfig(ctx context.Context, cli client.Client, contour *model.Contour) error {
+	current, err := current(ctx, cli, contour)
+
+	switch {
+	// Legitimate error: return it
+	case err != nil && !errors.IsNotFound(err):
+		return err
+	// ContourConfiguration not found: create it
+	case errors.IsNotFound(err):
+		contourConfig := &contour_api_v1alpha1.ContourConfiguration{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: contour.Namespace,
+				Name:      contour.ContourConfigurationName(),
+				Labels:    model.OwnerLabels(contour),
+			},
+			Spec: contour_api_v1alpha1.ContourConfigurationSpec{
+				Gateway: &contour_api_v1alpha1.GatewayConfig{
+					GatewayRef: &contour_api_v1alpha1.NamespacedName{
+						Namespace: contour.Namespace,
+						Name:      contour.Name,
+					},
+				},
+				Envoy: &contour_api_v1alpha1.EnvoyConfig{
+					Service: &contour_api_v1alpha1.NamespacedName{
+						Namespace: contour.Namespace,
+						Name:      contour.EnvoyServiceName(),
+					},
+				},
+			},
+		}
+
+		return cli.Create(ctx, contourConfig)
+	// Already exists: ensure it has the relevant fields set correctly.
+	default:
+		maybeUpdated := current.DeepCopy()
+		maybeUpdated.Spec.Gateway = &contour_api_v1alpha1.GatewayConfig{
+			GatewayRef: &contour_api_v1alpha1.NamespacedName{
+				Namespace: contour.Namespace,
+				Name:      contour.Name,
+			},
+		}
+
+		if maybeUpdated.Spec.Envoy == nil {
+			maybeUpdated.Spec.Envoy = &contour_api_v1alpha1.EnvoyConfig{}
+		}
+		maybeUpdated.Spec.Envoy.Service = &contour_api_v1alpha1.NamespacedName{
+			Namespace: contour.Namespace,
+			Name:      contour.EnvoyServiceName(),
+		}
+
+		if !equality.Semantic.DeepEqual(current, maybeUpdated) {
+			return cli.Update(ctx, maybeUpdated)
```

Comment:
i forget what the plan around updating and owner labels are but in the future for this and other resources, but a thing to solve in the future (e.g. if someone removes owner labels what do we do)

Reply:
Yeah, right now the existence of the owner labels is pretty essential to the provisioner controlling resources. There was some talk in the operator repo of switching to owner references, which could also work except there are some complications with cluster-scoped resources. IMHO it's OK to say that the resources must have the labels in order to be controlled by the provisioner; I do think the behavior if someone removes the labels from a managed resource may be weird, though (seems like the provisioner may try to create a new instance of the resource, which would fail because a resource with that name already exists). So probably an area for improvement.